compiler/pipes/calc-func-dep.cpp

```diff
@@ -76,30 +78,33 @@ VertexPtr CalcFuncDepPass::on_enter_vertex(VertexPtr vertex) {
       VarPtr to_var = other_function->param_ids[ii];
       if (to_var->is_reference) { //passed as reference
         while (val->type() == op_index) {
-          val = val.as<op_index>()->array();
+          val = val.as<op_index>()->array();    // from $a['b'] extract $a
         }
         VarPtr from_var = (val->type() == op_var ? val.as<op_var>()->var_id : val.as<op_instance_prop>()->var_id);
         if (from_var->is_in_global_scope()) {
-          data.global_ref_edges.emplace_back(from_var, to_var);
+          data.global_ref_edges.emplace_front(from_var, to_var);
         } else if (from_var->is_reference) {
-          data.ref_ref_edges.emplace_back(from_var, to_var);
+          data.ref_ref_edges.emplace_front(from_var, to_var);
         }
       }
     }
   } else if (auto ptr = vertex.try_as<op_func_ptr>()) {
     data.dep.push_back(ptr->func_id);
   } else if (auto var_vertex = vertex.try_as<op_var>()) {
     VarPtr var = var_vertex->var_id;
-    if (var->is_in_global_scope()) {
-      data.used_global_vars.push_back(var);
+    // here we add var for later ub check; we store only vars that can potentially lead to ub, see check-ub.cpp
+    if ((var->is_global_var() || var->is_class_static_var()) &&   // only globals (not even static inside functions)
```

Comment:
May it's worth to use `var->is_in_global_scope()` here?

Reply:
No, as we don't have to deal with function static vars

Reply:
But this logic is applied for static vars in recursion calls as well.

Reply:
Not exactly. Other functions can't have direct references to static var of current function. Recursion calls are supposed to be safe, as the context is known. `modified_global_vars` that we collect here are used only to find potential misbehavior of global references across different functions. BTW, including static vars in this analysis doesn't affect vkcom codegen at all (no SAFE_OP added), it only heavies execution.